core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/StackSyncService.java

```diff
@@ -0,0 +1,141 @@
+    private ResourceRepository resourceRepository;
+    @Inject
+    private HostGroupRepository hostGroupRepository;
+    @Inject
+    private HostMetadataRepository hostMetadataRepository;
+    @Inject
+    private AmbariClusterConnector ambariClusterConnector;
+    @Inject
+    private AmbariClientProvider ambariClientProvider;
+    @Inject
+    private AmbariHostsRemover ambariHostsRemover;
+    @javax.annotation.Resource
+    private Map<CloudPlatform, SyncStateFinder> syncStateFinders;
+
+    public void sync(Long stackId) throws Exception {
+        Stack stack = stackService.getById(stackId);
+        Set<InstanceMetaData> instanceMetaDatas = instanceMetaDataRepository.findAllWithTerminatedInStack(stackId);
+        String statusReason = "State of the cluster infrastructure has been synchronized.";
+        for (InstanceMetaData instanceMetaData : instanceMetaDatas) {
+            InstanceGroup instanceGroup = instanceMetaData.getInstanceGroup();
+            InstanceSyncState state = syncStateFinders.get(stack.cloudPlatform()).getState(stackId, instanceMetaData.getInstanceId());
+            if (DELETED.equals(state) && !instanceMetaData.isTerminated()) {
+                updateMetadataToTerminatedAndDeregisterFromAmbari(stackId, stack, instanceMetaData, instanceGroup);
+            } else if (RUNNING.equals(state) && !instanceMetaData.isRegistered()) {
+                updateMetaDataToRunning(stackId, instanceMetaData, instanceGroup);
+            } else if (STOPPED.equals(state) && !instanceMetaData.isTerminated()) {
+                updateMetaDataToTerminated(stackId, instanceMetaData, instanceGroup);
+            }
+        }
+        if (Status.stopStatusesForUpdate().contains(stack.getStatus())) {
+            stackUpdater.updateStackStatus(stack.getId(), STOPPED, statusReason);
+        } else if (Status.availableStatusesForUpdate().contains(stack.getStatus())) {
+            stackUpdater.updateStackStatus(stack.getId(), AVAILABLE, statusReason);
+        }
+
+
+    }
+
+    private void updateMetadataToTerminatedAndDeregisterFromAmbari(Long stackId, Stack stack, InstanceMetaData instanceMetaData, InstanceGroup instanceGroup) {
+        instanceGroup.setNodeCount(instanceGroup.getNodeCount() - 1);
+        instanceMetaData.setInstanceStatus(TERMINATED);
+        HostGroup hostGroup = hostGroupRepository.findHostGroupsByInstanceGroupName(stack.getCluster().getId(), instanceGroup.getGroupName());
+        HostMetadata data = hostMetadataRepository.findHostsInClusterByName(stack.getCluster().getId(), instanceMetaData.getDiscoveryFQDN());
+        try {
+            deregisterFromAmbari(stack, hostGroup, data);
+            instanceMetaDataRepository.save(instanceMetaData);
+            instanceGroupRepository.save(instanceGroup);
+            eventService.fireCloudbreakEvent(stackId, AVAILABLE.name(),
+                    String.format("Instance %s was terminated by hand update metadata...", instanceMetaData.getInstanceId()));
```

Comment:
it is not sure that it was terminated by hand - maybe it was terminated by Azure :)